rfcs/007-bytecode-interpreter.md

```diff
@@ -0,0 +1,839 @@
+implemented in OCaml's ZAM and Haskell's STG by having a proper `call`
+instruction. Compiling an expression in Tvix results in leaving some result on
+the stack. When evaluating an application `f x`, code is generated for `x`,
+followed by `f`, and finally a `CALL` opcode is generated. This is the
+eval-apply model, which should be less adapted to currying and partial
+application, at least without further optimizations.
+
+Looking at the implementation, it also seems that local stack slots are only
+deallocated after the whole code of the function has been run. In a lazy
+language any closure entry is a tail-call, so it looks like Tvix is retaining
+more memory than necessary[^tvix-no-documentation]: for example, in `let f = x:
+_continuation_; in let a = 1 + 1; in let b = a + 2; in f b`, `a` and `b` looks
+like they would remain alive on the stack for the whole duration of
+`_continuation_` although they aren't accessible anymore
+
+[^tvix-no-documentation]: Since the architecture of the Tvix VM isn't
+    technically documented, this is based on my understanding of the codebase in
+    a limited time; I could be missing something.
+
+#### References
+
+- [Tvix source code](https://github.com/tvlfyi/tvix)
+
+## Proposal
+
+The proposal takes inspiration from the previous VM descriptions, plus common
+and folklore knowledge about VM and interpreters in general. The design is
+guided by the following principles:
+
+- **fast compilation**: although we might provide facilities to distribute
+    compiled Nickel code in the future, this is just a bonus. The normal
+    workflow we want to optimize for is to compile Nickel code from source and
+    run it in a row. This must remain competitive with the current evaluation
+    model for small configurations with little to no computation. Thus, the
+    compiler must be fast (which often means it must be simple). Ideally, all
+    the program transformations and the compilation should be doable in one pass
+    (this might be hard to satisfy depending on current and future program
+    transformations, but it's an ideal goal).
+- **simplicity**: it goes hand in hand with fast compilation. But beyond
+    performance, we want to lay the foundation for a maintainable virtual
+    machine and thus start simple. We can leave many improvements and
+    optimizations for the future.
+- **optimize for speed**: in general, we favor optimizing for speed rather than
+    memory consumption, even if because of memory allocation, both are
+    correlated.
+
+### Virtual machine architecture
+
+The virtual machine is a stack-based machine with a set of working registers.
+
+It follows the same general principles as the ZAM, the Tvix VM or the STG
+machine. Those machines still have a number of differences on how they handle
+fundamental operations such as function application. We take inspiration from
+the
+[call-by-push-value](https://www.cs.bham.ac.uk/~pbl/papers/thesisqmwphd.pdf)(CBPV)
```

Comment:
:blush:

Reply:
*unevaluated thunk, rather

Reply:
Maybe I said that, I don't know. But I think my broader point was that I was thinking, at the time, of metadata (doc, default, …) as being attached to thunks. And CBPV helps us understand what's going on, because of thunking being an explicit operation in the language, so you can change what's going on there.

That being said, I always think of any evaluation model in terms of CBPV/polarised system L. Even when I don't present it that way: it's almost certainly a translation of the CPBV I have in my head. And I'm keen to embrace it in abstract machines as well.

Reply:
Speaking of, you can think of system L as representing an abstract machine with a structured stack. Implemented directly, it's not as cache friendly as an actual stack machine. But maybe it's a middle ground to consider, as you don't have to come up with a complex linearised continuation representation for pattern matching, in particular.

On the other hand, maybe it's too tree-like for this proposal. I don't know, I haven't thought about it. I just said the name, and this idea popped up in response. Do whatever you want to do with it.